File: addon/Data/Achievements.lua
local _, WS = ...

-- Achievements. `test(db, run, player)` is evaluated once per second during a
-- run and again when a run ends; `run`/`player` are nil outside a run.
-- Rewards: { type = "character"|"map", id = ... } or { type = "gold", amount = n }.
-- Everything here is checked by Achievements.lua and persisted per account.
WS.AchievementData = {
    first_blood = {
        name = "First Blood",
        description = "Slay your first enemy.",
        icon = "Interface\\Icons\\Ability_Rogue_Eviscerate",
        test = function(db) return db.records.totalKills >= 1 end,
    },
    take_his_candle = {
        name = "You Take Candle",
        description = "Defeat Foreman Grimtunnel.",
        icon = "Interface\\Icons\\INV_Misc_Bone_01",
        reward = { type = "gold", amount = 50 },
        test = function(db) return db.statistics.bossSlain.grimtunnel end,
    },
    gnollbane = {
        name = "Bane of the Riverpaw",
        description = "Defeat Gnarlfang the Ravager.",
        icon = "Interface\\Icons\\Ability_Hunter_Pet_Hyena",
        reward = { type = "gold", amount = 100 },
        test = function(db) return db.statistics.bossSlain.gnarlfang end,
    },
    unmasked = {
        name = "Unmasked",
        description = "Defeat the Masked Admiral.",
        icon = "Interface\\Icons\\Ability_Rogue_KidneyShot",
        reward = { type = "gold", amount = 250 },
        test = function(db) return db.statistics.bossSlain.masked_admiral end,
    },
    the_long_dark_ends = {
        name = "The Long Dark Ends",
        description = "Defeat Lich-Lord Marrowfrost.",
        icon = "Interface\\Icons\\Spell_Frost_FrostNova",
        reward = { type = "gold", amount = 1000 },
        test = function(db) return db.statistics.bossSlain.marrowfrost end,
    },

    -- Battlefield gates: surviving 10 minutes opens the next zone.
    beyond_the_forest = {
        name = "Beyond the Forest",
        description = "Survive for 10 minutes in Elwynn Forest.",
        icon = "Interface\\Icons\\INV_Misc_Flower_02",
        reward = { type = "map", id = "westfall" },
        rewardText = "Unlocks Westfall.",
        test = function(db, run) return run and run.mapId == "elwynn" and run.time >= 600 end,
    },
    into_the_dark = {
        name = "Into the Dark",
        description = "Survive for 10 minutes in Westfall.",
        icon = "Interface\\Icons\\INV_Misc_Food_Wheat_01",
        reward = { type = "map", id = "duskwood" },
        rewardText = "Unlocks Duskwood.",
        test = function(db, run) return run and run.mapId == "westfall" and run.time >= 600 end,
    },
    across_the_sea = {
        name = "Across the Great Sea",
        description = "Survive for 10 minutes in Duskwood.",
        icon = "Interface\\Icons\\INV_Misc_Bone_08",
        reward = { type = "map", id = "barrens" },
        rewardText = "Unlocks The Barrens.",
        test = function(db, run) return run and run.mapId == "duskwood" and run.time >= 600 end,
    },
    the_frozen_north = {
        name = "The Frozen North",
        description = "Survive for 10 minutes in The Barrens.",
        icon = "Interface\\Icons\\INV_Ammo_Snowball",
        reward = { type = "map", id = "icecrown" },
        rewardText = "Unlocks Icecrown.",
        test = function(db, run) return run and run.mapId == "barrens" and run.time >= 600 end,
    },

    -- Survivor unlocks.
    nightfall_survivor = {
        name = "Nightfall Survivor",
        description = "Survive for 8 minutes in a single run.",
        icon = "Interface\\Icons\\Ability_Rogue_FanofKnives",
        reward = { type = "character", id = "rogue" },
        rewardText = "Unlocks Dr. Rav McBreathless, the Rogue.",
        test = function(db, run) return run and run.time >= 480 end,
    },
    giant_slayer = {
        name = "Giant Slayer",
        description = "Slay 2 bosses in a single run.",
        icon = "Interface\\Icons\\Ability_Whirlwind",
        reward = { type = "character", id = "warrior" },
        rewardText = "Unlocks AAAAAAAAA, the Fallen Hero.",
        test = function(db, run) return run and run.bossesSlain >= 2 end,
    },
    seasoned_veteran = {
        name = "Seasoned Veteran",
        description = "Reach level 20 in a single run.",
        icon = "Interface\\Icons\\Ability_UpgradeMoonGlaive",
        reward = { type = "character", id = "hunter" },
        rewardText = "Unlocks Maeca Barefoot, the Hunter.",
        test = function(db, run, player) return player and player.level >= 20 end,
    },
    scourge_of_the_masses = {
        name = "Scourge of the Masses",
        description = "Slay 750 enemies across all runs.",
        icon = "Interface\\Icons\\Spell_Shadow_ShadowBolt",
        reward = { type = "character", id = "warlock" },
        rewardText = "Unlocks Nim B'ladin, the Warlock.",
        test = function(db) return db.records.totalKills >= 750 end,
    },
    fortune_seeker = {
        name = "Fortune Seeker",
        description = "Bank 500 gold across all runs.",
        icon = "Interface\\Icons\\INV_Misc_Coin_02",
        reward = { type = "character", id = "shaman" },
        rewardText = "Unlocks Vonnra Hydrocheck, the Shaman.",
        test = function(db) return db.statistics.goldEarned >= 500 end,
    },

    -- Feats.
    forbidden_knowledge = {
        name = "Forbidden Knowledge",
        description = "Evolve a weapon to its ultimate form.",
        icon = "Interface\\Icons\\Spell_Shadow_DarkRitual",
        reward = { type = "gold", amount = 150 },
        test = function(db) return db.statistics.evolutions >= 1 end,
    },
    lights_favor = {
        name = "Light's Favor",
        description = "Survive for 3 minutes straight without taking damage.",
        icon = "Interface\\Icons\\Spell_Holy_Devotion",
        reward = { type = "gold", amount = 150 },
        test = function(db, run) return run and run.noHitStreak >= 180 end,
    },
    walking_armory = {
        name = "Walking Armory",
        description = "Wield 6 weapons at once.",
        icon = "Interface\\Icons\\INV_Misc_EngGizmos_17",
        reward = { type = "gold", amount = 150 },
        test = function(db, run, player) return player and #player.weapons >= 6 end,
    },
    mrglglgl = {
        name = "Mrglglglgl!",
        description = "Slay 500 murlocs across all runs.",
        icon = "Interface\\Icons\\INV_Misc_Fish_13",
        reward = { type = "gold", amount = 200 },
        test = function(db) return (db.statistics.killsByFamily.murloc or 0) >= 500 end,
    },
    master_craftsman = {
        name = "Master Craftsman",
        description = "Forge a weapon union.",
        icon = "Interface\\Icons\\Ability_Warrior_PunishingBlow",
        reward = { type = "gold", amount = 250 },
        test = function(db) return db.statistics.unions and db.statistics.unions > 0 end,
    },
    grave_robber = {
        name = "Grave Robber",
        description = "Open a weathered coffin found on the battlefield.",
        icon = "Interface\\Icons\\INV_Misc_Urn_01",
        reward = { type = "gold", amount = 200 },
        test = function(db) return db.statistics.coffinsOpened and db.statistics.coffinsOpened > 0 end,
    },
    the_light_curdles = {
        name = "The Light Curdles",
        description = "Claim a runeblade drawn out by your own desecration.",
        icon = "Interface\\Icons\\Spell_DeathKnight_BloodPresence",
        reward = { type = "gold", amount = 400 },
        test = function(db) return (db.statistics.runebladesClaimed or 0) > 0 end,
    },
    you_are_prepared = {
        name = "You Are Prepared",
        description = "Claim the warglaives after giving yourself to the fel.",
        icon = "Interface\\Icons\\Spell_Shadow_Metamorphosis",
        reward = { type = "gold", amount = 400 },
        test = function(db) return (db.statistics.warglaivesClaimed or 0) > 0 end,
    },
    survivor_of_the_long_dark = {
        name = "Survivor of the Long Dark",
        description = "Achieve Victory on any battlefield.",
        icon = "Interface\\Icons\\Spell_Holy_AuraOfLight",
        reward = { type = "gold", amount = 500 },
        test = function(db) return db.statistics.totalVictories >= 1 end,
    },
}

WS.AchievementOrder = {
    "first_blood", "take_his_candle", "gnollbane", "unmasked", "the_long_dark_ends",
    "beyond_the_forest", "into_the_dark", "across_the_sea", "the_frozen_north",
    "nightfall_survivor", "giant_slayer", "seasoned_veteran", "scourge_of_the_masses",
    "fortune_seeker", "forbidden_knowledge", "lights_favor", "walking_armory",
    "mrglglgl", "master_craftsman", "grave_robber", "survivor_of_the_long_dark",
}

File: src/data/achievements.js
/* Achievements. `test(db)` reads persisted statistics; `reward` grants gold or
 * unlocks a character/map the moment the test first passes. */
'use strict';
(function (WS) {

  WS.Achievements = {
    first_blood: {
      name: 'First Blood', art: 'blade',
      description: 'Slay your first enemy.',
      test: (db) => db.statistics.totalKills >= 1,
    },
    take_his_candle: {
      name: 'You Take Candle', art: 'candle',
      description: 'Defeat Foreman Grimtunnel.',
      reward: { type: 'gold', amount: 50 },
      test: (db) => !!db.statistics.bosses.grimtunnel,
    },
    gnollbane: {
      name: 'Bane of the Riverpaw', art: 'claw',
      description: 'Defeat Gnarlfang the Ravager.',
      reward: { type: 'gold', amount: 100 },
      test: (db) => !!db.statistics.bosses.gnarlfang,
    },
    unmasked: {
      name: 'Unmasked', art: 'mask',
      description: 'Defeat the Masked Admiral.',
      reward: { type: 'gold', amount: 250 },
      test: (db) => !!db.statistics.bosses.masked_admiral,
    },
    the_long_dark_ends: {
      name: 'The Long Dark Ends', art: 'frostaura',
      description: 'Defeat Lich-Lord Marrowfrost.',
      reward: { type: 'gold', amount: 1000 },
      test: (db) => !!db.statistics.bosses.marrowfrost,
    },

    /* ------------------------------------------------- map progression --- */
    beyond_the_forest: {
      name: 'Beyond the Forest', art: 'leaf',
      description: 'Survive for 10 minutes in Elwynn Forest.',
      reward: { type: 'map', id: 'westfall' },
      test: (db) => (db.statistics.bestTime.elwynn || 0) >= 600,
    },
    into_the_dark: {
      name: 'Into the Dark', art: 'deadtree',
      description: 'Survive for 10 minutes in Westfall.',
      reward: { type: 'map', id: 'duskwood' },
      test: (db) => (db.statistics.bestTime.westfall || 0) >= 600,
    },
    across_the_sea: {
      name: 'Across the Great Sea', art: 'sun',
      description: 'Survive for 10 minutes in Duskwood.',
      reward: { type: 'map', id: 'barrens' },
      test: (db) => (db.statistics.bestTime.duskwood || 0) >= 600,
    },
    the_frozen_north: {
      name: 'The Frozen North', art: 'crystal',
      description: 'Survive for 10 minutes in The Barrens.',
      reward: { type: 'map', id: 'icecrown' },
      test: (db) => (db.statistics.bestTime.barrens || 0) >= 600,
    },

    /* ------------------------------------------------ survivor unlocks --- */
    nightfall_survivor: {
      name: 'Nightfall Survivor', art: 'moon',
      description: 'Survive for 8 minutes in a single run.',
      reward: { type: 'character', id: 'rogue' },
      test: (db) => db.statistics.bestRunTime >= 480,
    },
    giant_slayer: {
      name: 'Giant Slayer', art: 'skull',
      description: 'Slay 2 bosses in a single run.',
      reward: { type: 'character', id: 'warrior' },
      test: (db) => db.statistics.bestBossesInRun >= 2,
    },
    seasoned_veteran: {
      name: 'Seasoned Veteran', art: 'book',
      description: 'Reach level 20 in a single run.',
      reward: { type: 'character', id: 'hunter' },
      test: (db) => db.statistics.bestLevel >= 20,
    },
    scourge_of_the_masses: {
      name: 'Scourge of the Masses', art: 'blade',
      description: 'Slay 750 enemies across all runs.',
      reward: { type: 'character', id: 'warlock' },
      test: (db) => db.statistics.totalKills >= 750,
    },
    fortune_seeker: {
      name: 'Fortune Seeker', art: 'coin',
      description: 'Bank 500 gold across all runs.',
      reward: { type: 'character', id: 'shaman' },
      test: (db) => db.statistics.totalGold >= 500,
    },

    /* ------------------------------------------------------- challenge --- */
    forbidden_knowledge: {
      name: 'Forbidden Knowledge', art: 'arcane',
      description: 'Evolve a weapon to its ultimate form.',
      reward: { type: 'gold', amount: 150 },
      test: (db) => db.statistics.evolutions >= 1,
    },
    lights_favor: {
      name: "Light's Favor", art: 'aegis',
      description: 'Survive for 3 minutes straight without taking damage.',
      reward: { type: 'gold', amount: 150 },
      test: (db) => db.statistics.bestNoHitStreak >= 180,
    },
    walking_armory: {
      name: 'Walking Armory', art: 'sword',
      description: 'Wield 6 weapons at once.',
      reward: { type: 'gold', amount: 150 },
      test: (db) => db.statistics.maxWeapons >= 6,
    },
    mrglglgl: {
      name: 'Mrglglglgl!', art: 'murloc',
      description: 'Slay 500 murlocs across all runs.',
      reward: { type: 'gold', amount: 200 },
      test: (db) => (db.statistics.families.murloc || 0) >= 500,
    },
    master_craftsman: {
      name: 'Master Craftsman', art: 'anvil',
      description: 'Forge a weapon union.',
      reward: { type: 'gold', amount: 250 },
      test: (db) => db.statistics.unions >= 1,
    },
    grave_robber: {
      name: 'Grave Robber', art: 'grave',
      description: 'Open a weathered coffin found on the battlefield.',
      reward: { type: 'gold', amount: 200 },
      test: (db) => (db.statistics.coffinsOpened || 0) > 0,
    },
    the_light_curdles: {
      name: 'The Light Curdles', art: 'desecrate',
      description: 'Claim a runeblade drawn out by your own desecration.',
      reward: { type: 'gold', amount: 400 },
      test: (db) => (db.statistics.runebladesClaimed || 0) > 0,
    },
    you_are_prepared: {
      name: 'You Are Prepared', art: 'soulrend',
      description: 'Claim the warglaives after giving yourself to the fel.',
      reward: { type: 'gold', amount: 400 },
      test: (db) => (db.statistics.warglaivesClaimed || 0) > 0,
    },
    survivor_of_the_long_dark: {
      name: 'Survivor of the Long Dark', art: 'sun',
      description: 'Achieve Victory on any battlefield.',
      reward: { type: 'gold', amount: 500 },
      test: (db) => db.statistics.totalVictories >= 1,
    },
    eclipse_broken: {
      name: 'Eclipse Broken', art: 'sovereign',
      description: 'Defeat Aethelgard in the Eclipse Arena.',
      reward: { type: 'gold', amount: 1500 },
      test: (db) => !!db.statistics.bosses.aethelgard,
    },
  };

  WS.AchievementOrder = [
    'first_blood', 'take_his_candle', 'gnollbane', 'unmasked', 'the_long_dark_ends',
    'beyond_the_forest', 'into_the_dark', 'across_the_sea', 'the_frozen_north',
    'nightfall_survivor', 'giant_slayer', 'seasoned_veteran', 'scourge_of_the_masses',
    'fortune_seeker', 'forbidden_knowledge', 'lights_favor', 'walking_armory',
    'mrglglgl', 'master_craftsman', 'grave_robber', 'the_light_curdles',
    'you_are_prepared', 'survivor_of_the_long_dark', 'eclipse_broken',
  ];

})(window.WS);
